### PR TITLE
`Extractor`: implement `usage` and `installation` schemas

### LINKED ISSUE #19: `Extractor:attibutes:usage` needs documentation and schema
During the last office hours, we have settled on a first draft for the two variants of `usage`:

- `usage: "cli: executable {{ file_type }} {{ file_path }} {{ output_file }}"` for executing an `Extractor` from shell, with the obvious meanings of the parameters, returning the data as a file, and
- `usage: "python: project.module.function( {{ file_type }, {{ file_path }} )"` for executing an `Extractor` from python, returning the data as an in-memory object

This functionality is currently:
- undocumented: there is no explanation in the `Extractor` schema on how to compose these `usage` strings,
- unvalidated: the `Registry` machinery does not have a way to validate the `usage` string for syntax-correctness
- inconsistent: I would suggest renaming the `{{ file_path }}` pattern to `{{ input_file }}` for consistency with `{{ output_file }}`

### PR DESCRIPTION
**Pull request by PeterKraus**
This PR re-implements the `usage` and `installation` attributes of an `Extractor`, so that the execution via the api can be fully self-contained using information from within the registry. Closes #19 .

- [x] `usage` reimplemented:
- [x] `usage:method` child entry uses an `enum` (currently `cli` and `python` are the only options)
- [x] `usage:setup` child entry for defining e.g. `import extractor` or `import package.extractor` statements
- [x] `usage:command` child entry has a defined set of template words (e.g. `$filetype`) and validates its syntax
- [x] `installation` implemented:
- [x] `installation:method` child entry uses an `enum` (currently `pip` and `conda` are the only options)
- [x] `installation:requirements` child `str` entry optionally contains contents of an "requirements.txt" or yaml file
- [x] `installation:packages` child `list[str]` entry containing packages (versioned, git+https, etc.) to be installed by `installation:method`

### CODE REVIEW

File: schemas/extractor.yml [hunk 1]
```diff
@@ -27,6 +27,44 @@ classes:
...
+            method:
+                required: true
+                range: InstallationTypes
+                description: >-
+                    Extractor installation method, specifying the package manager
+                    to
+                    process the installation instructions with.
+            requirements:
+                description: >-
+                    The contents of an optional 'requirements.yml'-like file, which
```
Comment (ml-evs) [inline schemas/extractor.yml]:
```suggestion
                    The contents of an optional 'requirements.txt'-like file, which
```

is this a new thing or should it be .txt?

File: examples/extractor/example_extractor.yml [hunk 2]
```diff
@@ -16,10 +16,13 @@ source_repository: >-
...
-    - cli:parse --type=example {{ file_path }}
-    - python:example.extrators.extract({{ file_path }}, type="example")
-    - python:example.extrators.extract.ExampleExtractor({{ file_path }}).parse()
+usage:
+    - method: cli
+      command: parse --type=example {{ file_path }}
+    - method: python
+      command: example.extrators.extract({{ file_path }}, type="example")
+    - method: python
+      command: example.extrators.extract.ExampleExtractor({{ file_path }}).parse()
```
Comment (ml-evs) [inline examples/extractor/example_extractor.yml]:
```suggestion
      command: example.extractors.extract({{ file_path }}, type="example")
    - method: python
      command: example.extractors.extract.ExampleExtractor({{ file_path }}).parse()
```

File: examples/extractor/example_extractor.yml [hunk 3]
```diff
@@ -16,10 +16,13 @@ source_repository: >-
...
-usage: >-
-    - cli:parse --type=example {{ file_path }}
-    - python:example.extrators.extract({{ file_path }}, type="example")
-    - python:example.extrators.extract.ExampleExtractor({{ file_path }}).parse()
+usage:
+    - method: cli
+      command: parse --type=example {{ file_path }}
+    - method: python
+      command: example.extrators.extract({{ file_path }}, type="example")
+    - method: python
```
Comment (ml-evs) [inline examples/extractor/example_extractor.yml:25]:
```suggestion
    - method: python
      import: example.extractors.extract
```

File: schemas/extractor.yml [hunk 4]
```diff
@@ -68,9 +106,55 @@ classes:
...
+                    using arguments, shall be supplied in the 'command' slot.
+
+
+    InstallationTypes:
+        permissible_values:
+            pip:
+                description: >-
+                    Install the Extractor using 'pip'. The 'command' has to be a valid
+                    pip invocation, such as 'pip install yadg'. Usage of a requirements
+                    file can be specified as 'pip install -r $requirements'. Packages
```
Comment (ml-evs) [inline schemas/extractor.yml]:
```suggestion
                    file can be specified as 'pip install $requirements'. Packages
```

I think?

File: schemas/extractor.yml [hunk 5]
```diff
@@ -27,6 +27,44 @@ classes:
...
+                    process the installation instructions with.
+            requirements:
+                description: >-
+                    The contents of an optional 'requirements.txt'-like file, which
+                    can be forwarded to the installation command using '$requirements'.
+            command:
+                required: true
+                description: >-
+                    A machine-executable command by which the Extractor can be installed
+                    using the 'method' specified above.
```
Comment (PeterKraus) [inline schemas/extractor.yml]: Rename to `packages` and allow a `list[str]` here which get `pip install`-ed (can be versioned or git+https://...)

File: schemas/extractor.yml [hunk 6]
```diff
@@ -27,6 +27,44 @@ classes:
...
+                required: true
+                range: InstallationTypes
+                description: >-
+                    Extractor installation method, specifying the package manager
+                    to
+                    process the installation instructions with.
+            requirements:
+                description: >-
+                    The contents of an optional 'requirements.txt'-like file, which
+                    can be forwarded to the installation command using '$requirements'.
```
Comment (PeterKraus) [inline schemas/extractor.yml]: `requirements` and `packages` should be either or.

File: schemas/extractor.yml [hunk 7]
```diff
@@ -27,6 +27,44 @@ classes:
                     to this file type.
 
 
+    Usage:
+        attributes:
```
Comment (PeterKraus) [inline schemas/extractor.yml:31]: Add `import` keyword here, if possible figure out a way to default it to the `Extractor:id`

Comment (PeterKraus) [reply schemas/extractor.yml:31]: Cannot use `import` because of python, using `setup` instead.

File: examples/extractor/example_extractor.yml [hunk 8]
```diff
@@ -46,4 +64,4 @@ supported_filetypes:
       description: >-
           Can parse biologic MPR provided its a full moon.
 license:
-    spdx: MIT
+    uri: https://example.com
```
Comment (PeterKraus) [inline examples/extractor/example_extractor.yml:67]: Changed to check both `enums`.

**Comment by PeterKraus** [comment]
@ml-evs this should be in a better shape now, please have a look once you have time. One thing that's still open is the set / format of "keywords", related to the `Extractor:usage:command` field. We can either push this PR as is and leave #19 open, or resolve #19 there and then add a longer docstring for that attribute here.

File: schemas/extractor.yml [hunk 9]
```diff
@@ -27,6 +27,50 @@ classes:
...
+                      - for 'method=python', this is an argument to python's 'import'
+            command:
+                required: true
+                description: >-
+                    A machine-executable command by which the Extractor functionality
+                    can be accessed:
+                      - for 'method=cli' a parametrized executable invocation which
+                        should be available after installing the Extractor;
+                      - for 'method=python' a parametrized function invocation which
+                        should eb available after installing and setting up the Extractor.
```
Comment (ml-evs) [inline schemas/extractor.yml]:
```suggestion
                        should be available after installing and setting up the Extractor.
```

File: schemas/extractor.yml [hunk 10]
```diff
@@ -27,6 +27,50 @@ classes:
...
+                description: >-
+                    Extractor installer, specifying the package manager to process
+                    the installation instructions with.
+            requirements:
+                description: >-
+                    Contents of a 'requirements.txt'-like file. Will be installed
+                    by the
+                    selected Installer using an appropriate method, e.g. 'pip install
+                    -r requirements.txt' for pip, or 'conda env create -f requirements.txt'
+                    for conda.
```
Comment (ml-evs) [inline schemas/extractor.yml]:
```suggestion
                    Contents of a 'requirements.txt'-like file. Will be installed
                    by the selected Installer using an appropriate method, e.g.,
                    `pip install -r requirements.txt` for pip, or 
                    `conda env create -f requirements.txt` for conda.
```

Going to reformat some of these single quoted strings with backticks as they will be rendered as monospace in the API docs

File: schemas/extractor.yml [hunk 11]
```diff
@@ -27,6 +27,50 @@ classes:
...
+            packages:
+                multivalued: true
+                description: >-
+                    A list of packages, including versioned packages or git+https://
```
Comment (ml-evs) [inline schemas/extractor.yml]:
```suggestion
                    A list of packages, including versioned packages or `git+https://`
```

File: examples/extractor/example_extractor.yml [hunk 12]
```diff
@@ -16,10 +16,28 @@ source_repository: >-
...
+            - python=3.4
+            - numpy
+            - toolz
+            - matplotlib
+            - dill
+            - pandas
+            - partd
+            - bokeh
+            - pip:
+              - git+https://github.com/blaze/dask.git#egg=dask[complete]
```
Comment (ml-evs) [inline examples/extractor/example_extractor.yml:40]: Looking at the instructions below, would the equivalent `pip` example here just be:

```yaml
- method: pip
  packages: example-extractor
```

? I guess my only confusion would be that you could also just inline `example-extractor` as a requirement, so not sure I see the difference. I guess its really that one is intended to provide a full reproducible environment, in which case perhaps `environment` is a better name?

File: schemas/extractor.yml (hunk 9 above)
Comment (ml-evs) [inline schemas/extractor.yml]:
```suggestion
                      - for `method: cli` a parametrized executable invocation which
                        should be available after installing the Extractor;
                      - for `method: python` a parametrized function invocation which
                        should be available after installing and setting up the Extractor.
```

File: schemas/extractor.yml [hunk 13]
```diff
@@ -27,6 +27,50 @@ classes:
...
+        attributes:
+            method:
+                required: true
+                range: UsageTypes
+                description: >-
+                    Usage invocation method, e.g. from a command line or from Python.
+            setup:
+                description: >-
+                    Any necessary setup step for the 'command' to become available:
+                      - for 'method=python', this is an argument to python's 'import'
```
Comment (ml-evs) [inline schemas/extractor.yml]:
```suggestion
                      - for `method: python`, this is an argument to Python's `import`
```

File: schemas/extractor.yml [hunk 14]
```diff
@@ -27,6 +27,50 @@ classes:
...
+            command:
+                required: true
+                description: >-
+                    A machine-executable command by which the Extractor functionality
+                    can be accessed:
+                      - for 'method=cli' a parametrized executable invocation which
+                        should be available after installing the Extractor;
+                      - for 'method=python' a parametrized function invocation which
```
Comment (ml-evs) [inline schemas/extractor.yml]:
```suggestion
                      - for `method: cli` a parametrized executable invocation which
                        should be available after installing the Extractor;
                      - for `method: python` a parametrized function invocation which
```

File: schemas/extractor.yml [hunk 15]
```diff
@@ -65,9 +109,49 @@ classes:
...
             usage:
                 multivalued: true
+                inlined_as_list: true
+                range: Usage
                 description: >-
-                    A command-line invocation of the parser, to be available after
-                    installation instructions have been followed.
+                    A machine-actionable description of the usage of the Extractor.
+                    The described usage pattern shall be available after the
+                    'installation' instructions have been followed.
```
Comment (ml-evs) [inline schemas/extractor.yml]:
```suggestion
                    `installation` instructions have been followed.
```

File: schemas/extractor.yml [hunk 16]
```diff
@@ -65,9 +109,49 @@ classes:
...
+                description: >-
+                    A machine-actionable set of installation instructions to obtain
+                    a working set-up of the Extractor.
             instructions:
                 description: >-
-                    Any usage or installation instructions for this extractor.
+                    Any usage notes or installation instructions for this Extractor.
+                    This field is intended for human use only and is not intended
+                    to be machine-actionable. Please use the 'installation' and
+                    'usage' fields for that purpose.
```
Comment (ml-evs) [inline schemas/extractor.yml]:
```suggestion
                    to be machine-actionable. Please use the `installation` and
                    `usage` fields for that purpose.
```

File: schemas/extractor.yml [hunk 17]
```diff
@@ -65,9 +109,49 @@ classes:
...
+
+
+enums:
+    UsageTypes:
+        permissible_values:
+            cli:
+                description: >-
+                    Access the Extractor as a command line executable. The executable
+                    as well as the invocation syntax, parametrised using arguments,
+                    shall be supplied in the 'command' slot.
```
Comment (ml-evs) [inline schemas/extractor.yml]:
```suggestion
                    shall be supplied in the `command` slot.
```

File: schemas/extractor.yml [hunk 18]
```diff
@@ -65,9 +109,49 @@ classes:
...
+            python:
+                description: >-
+                    Access the Extractor as a Python function. The function should
+                    be available when 'import Extractor' is called, and should be
```
Comment (ml-evs) [inline schemas/extractor.yml]:
```suggestion
                    be available when `import Extractor` is called, and should be
```

I still think this is a bit unclear, I guess now this should be `import <setup>`?

File: schemas/extractor.yml [hunk 19]
```diff
@@ -65,9 +109,49 @@ classes:
...
+
+    InstallerTypes:
+        permissible_values:
+            pip:
+                description: >-
+                    Install the Extractor using 'pip'. Will process 'packages' using
+                    'pip install' and 'requirements' using 'pip install -r'.
+            conda:
+                description: >-
+                    Install the Extractor using 'conda'. Not yet implemented / supported.
```
Comment (ml-evs) [inline schemas/extractor.yml]:
```suggestion
                    Install the Extractor using 'pip'. Will process `packages` using
                    `pip install` and `requirements` using `pip install -r`.
            conda:
                description: >-
                    Install the Extractor using 'conda'. Not yet implemented / supported.
```

**Review by ml-evs — COMMENTED** [review]
Mostly formatting changes, I think I will self-commit all those to reduce noise (will check on slack first) to leave the minor outstanding comments

File: examples/extractor/example_extractor.yml (hunk 3 above)
Comment (ml-evs) [reply examples/extractor/example_extractor.yml:25]: Resolved by the addition of `setup`

File: examples/extractor/example_extractor.yml (hunk 12 above)
Comment (PeterKraus) [reply examples/extractor/example_extractor.yml:40]: I have just lifted that one from [here](https://github.com/binder-project/example-conda-environment/blob/master/environment.yml)...

File: examples/extractor/example_extractor.yml [hunk 20]
```diff
@@ -16,10 +16,28 @@ source_repository: >-
...
-usage: >-
-    - cli:parse --type=example {{ file_path }}
-    - python:example.extrators.extract({{ file_path }}, type="example")
-    - python:example.extrators.extract.ExampleExtractor({{ file_path }}).parse()
+usage:
+    - method: cli
+      command: parse --type=example {{ file_path }}
+    - method: python
+      setup: example.extractors
+      command: extractors.extract({{ file_path }}, type="example")
```
Comment (ml-evs) [inline examples/extractor/example_extractor.yml]:
```suggestion
      command: example.extractors.extract({{ file_path }}, type="example")
```

I believe?

I guess we could also allow syntax like:

- `example.extractors as example`
- `ExampleExtractor from example.extractors`
- `* from example.extractors`

Comment (PeterKraus) [reply examples/extractor/example_extractor.yml]: I think the original is correct (without `example.` given the `setup:` field above).

As for the second point, I think we'll be using `importilb.import_module()`, which means we're restricted to importing modules/packages. We will still need to figure out how to excise the function name out of the `command:`; it's perhaps better to not complicate things and allow only modules into `setup:`.

Comment (ml-evs) [reply examples/extractor/example_extractor.yml]: This is what I'm getting at for the first, replace `setup: scipy.signal` and command `signal.savgol_filter` and you get this:

```
>>> import scipy.signal
>>> signal.savgol_filter
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
NameError: name 'signal' is not defined
>>> scipy.signal.savgol_filter
<function savgol_filter at 0x7f9bafc1d0d0>
```

or should I be reading it as `setup: scipy.signal` and `command: savgol_filter` needing to then be composed to `scipy.signal.savgol_filter`? In that case I think its just the plural that's incorrect:

```suggestion
      command: extractor.extract({{ file_path }}, type="example")
```

Comment (ml-evs) [reply examples/extractor/example_extractor.yml]: Fine for the second point though!

Comment (PeterKraus) [reply examples/extractor/example_extractor.yml]: You're right. I'll commit the first suggestion.

File: schemas/extractor.yml [hunk 21]
```diff
@@ -27,6 +27,49 @@ classes:
...
+                      - for `method: python`, this is an argument to Python's `import`
+            command:
+                required: true
+                description: >-
+                    A machine-executable command by which the Extractor functionality
+                    can be accessed:
+                      - for `method: cli` a parametrized executable invocation which
+                        should be available after installing the Extractor;
+                      - for `method: python` a parametrized function invocation which
+                        should be available after installing and setting up the Extractor.
```
Comment (ml-evs) [inline schemas/extractor.yml:49]:
```suggestion
                      - for `method: python` a parametrized function invocation which
                        should be available after installing and setting up the Extractor.
                        
                    The following templated parameters can be requested from the user, using the `{{}}`` syntax:
                      - `{{ input_path }}`: the location of the resource (e.g., file, or directory) on disk to parse (REQUIRED).
                      - `{{ input_type }}`: the registry ID of the file type to parse (OPTIONAL).
                      - `{{ output_path }}`: the location of the output to write to disk with the results of the extraction, if successful (OPTIONAL).
                      - `{{ output_type }}`: the registry ID of the file type to write to disk (OPTIONAL).
```

File: schemas/extractor.yml (hunk 18 above)
Comment (ml-evs) [reply schemas/extractor.yml]:
```suggestion
                    be available when `import <setup>` is called, and should be
```

File: schemas/extractor.yml [hunk 22]
```diff
@@ -65,9 +108,69 @@ classes:
...
+                description: >-
+                    Access the Extractor as a command line executable. The executable
+                    as well as the invocation syntax, parametrised using arguments,
+                    shall be supplied in the `command` slot.
+            python:
+                description: >-
+                    Access the Extractor as a Python function. The function should
+                    be available when 'import Extractor' is called, and should be
+                    specified including relevant namespaces. The function signature,
+                    parametrised using arguments, shall be supplied in the 'command'
```
Comment (ml-evs) [inline schemas/extractor.yml]:
```suggestion
                    parametrised using arguments, shall be supplied in the `command`
```

File: schemas/extractor.yml [hunk 23]
```diff
@@ -27,6 +27,60 @@ classes:
...
+                description: >-
+                    A machine-executable command by which the Extractor functionality
+                    can be accessed:
+                      - for `method: cli` a parametrized executable invocation which
+                        should be available after installing the Extractor;
+                      - for `method: python` a parametrized function invocation which
+                        should be available after installing and setting up the Extractor.
+
+                    The following templated parameters can be requested from the user,
+                    using the `{{}}`` syntax:
```
Comment (ml-evs) [inline schemas/extractor.yml]:
```suggestion
                    using the `{{}}` syntax:
```

File: schemas/extractor.yml [hunk 24]
```diff
@@ -50,7 +104,25 @@ classes:
                 required: true
                 range: SupportedFileTypes
                 description: >-
-                    Container for links to the file types that this extractor supports.
+                    Container for links to the file types that this extractor supports,
+                    matching the registry file types. The ID values can be passed
+                    to an extractor with the `{{ intput_type }}` parameter in `usage->command`.
```
Comment (ml-evs) [inline schemas/extractor.yml]:
```suggestion
                    to an extractor with the `{{ input_type }}` parameter in `usage->command`.
```

File: schemas/extractor.yml [hunk 25]
```diff
@@ -50,7 +104,25 @@ classes:
...
+                    Container for links to the file types that this extractor supports,
+                    matching the registry file types. The ID values can be passed
+                    to an extractor with the `{{ input_type }}` parameter in `usage->command`.
+
+            supported_output_filetypes:
+                multivalued: true
+                required: false
+                range: SupportedFileTypes
+                description: >-
+                    Countainer for enumerating the possible output formats of this
```
Comment (ml-evs) [inline schemas/extractor.yml]:
```suggestion
                    Container for enumerating the possible output formats of this
```

**Review by ml-evs — APPROVED** [review]
🚀

**Pull request merged**